### PR TITLE
feat (core): user session

### PR DESCRIPTION
**Pull request by MoumitaM**
## Description of the change

> Description of the change can be found [here](https://www.notion.so/rudderstacks/User-Session-Specs-JS-1c25f7f907d54eca8052abef16ec2d1d?d=97ba05206e7a45f1bece69b901a08819#1242fbf7e12b4e3785c9755920ef9e9b)

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

- - -
This change is [Reviewable](https://reviewable.io/reviews/rudderlabs/rudder-sdk-js/637)

### CODE REVIEW

**Comment by saikumarrs** [comment]
Please change the PR title to conventional commits format.

File: analytics.js [hunk 1]
```diff
@@ -1351,6 +1384,22 @@ class Analytics {
 
     return returnObj;
   }
+
+  /**
+   * A public method to start a session
+   * @param {string} sessionId     session identifier
+   * @returns
+   */
+  startSession(sessionId) {
```
Comment (pallabmaiti) [inline analytics.js:1393]: Did we consider this?
// in this case we'll calculate the sessionId
rudderClient.startSession();

Comment (MoumitaM) [reply analytics.js:1393]: Yes, in this function sessionId is optional.

File: analytics.js [hunk 2]
```diff
@@ -813,6 +815,21 @@ class Analytics {
...
+        try {
+          const timestamp = Date.now();
+          rudderElement.message.context.sessionId = this.session.getSessionId(timestamp);
+          if (!this.session.sessionInfo.sessionStart)
+            rudderElement.message.context.sessionStart = true;
+          this.session.update(timestamp);
+        } catch (e) {
+          handleError(e);
+        }
+      }
```
Comment (saikumarrs) [inline analytics.js:832]: This logic needs to be updated as discussed.

File: analytics.js [hunk 3]
```diff
@@ -983,6 +1000,14 @@ class Analytics {
...
     this.groupId = "";
     this.groupTraits = {};
+    const { manuallyTrackSession } = this.session.sessionInfo;
+    this.session.sessionInfo = {};
+    if (this.session.autoTrack) {
+      this.session.startAutoTracking();
+    }
+    if (manuallyTrackSession) {
+      this.session.start();
+    }
```
Comment (saikumarrs) [inline analytics.js:1010]: Move all of this into the session's reset API.

File: analytics.js [hunk 4]
```diff
@@ -1105,6 +1130,14 @@ class Analytics {
         this.sendAdblockPageOptions = options.sendAdblockPageOptions;
       }
     }
+
+    // Session initialization
+    try {
```
Comment (saikumarrs) [inline analytics.js:1135]: Move the exception handling into the session class.

**Review by saikumarrs — CHANGES_REQUESTED** [review]

File: session/index.js [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+  /**
+   * A function to validate current session and return true/false depending on that
+   * @param {number} timestamp
+   * @returns boolean
+   */
+  isValidSession(timestamp) {
+    return (
+      Object.keys(this.sessionInfo).length &&
+      typeof this.sessionInfo.expiresAt === 'number' &&
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply session/index.js:78]: I checked the implementation again. It seems there is no point in moving it to initialization. We can remove this.

File: session/index.js [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+      // and will proceed with it
+      if (timeout > 0 && timeout < MIN_SESSION_TIMEOUT) {
+        logger.warn('[Session]:: It is not advised to set "timeout" less than 10 seconds');
+      }
+      this.timeout = timeout;
+    }
+    // If auto session tracking is enabled start the session tracking
+    if (this.autoTrack) {
+      this.startAutoTracking();
+    } else if (
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply session/index.js]: Added.

File: session/index.js [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+    /**
+     * By default this.autoTrack will be true
+     * Cases where this.autoTrack will be false:
+     * 1. User explicitly set autoTrack load option to false
+     * 2. When user is manually tracking the session
+     *
+     * Depending on the use case, this.autoTrack is set to true/false.
+     */
+    this.autoTrack = !(
+      (options && options.sessions && options.sessions.autoTrack === false) ||
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply session/index.js]: Done.

File: session/index.js [hunk 8]
```diff
@@ -0,0 +1,155 @@
+/* eslint-disable import/prefer-default-export */
+import Storage from '../utils/storage';
+import logger from '../utils/logUtil';
+import {
+  DEFAULT_SESSION_TIMEOUT,
+  MIN_SESSION_TIMEOUT,
+} from "../utils/constants";
+
+class Session {
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply session/index.js]: Done.

File: session/index.js [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+    this.storage.setSessionInfo(this.sessionInfo);
+  }
+
+  /**
+   * A public method to start a session
+   * @param {string} sessionId     session identifier
+   * @returns
+   */
+  start(sessionId) {
+    if (sessionId && typeof sessionId !== 'string') {
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply session/index.js:111]: Done.

File: session/index.js [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+      return;
+    }
+    if (
+      !this.sessionInfo.manuallyTrackSession ||
+      !Object.keys(this.sessionInfo).length ||
+      !this.sessionInfo.id
+    ) {
+      this.autoTrack = false;
+      this.sessionInfo = {
+        id: sessionId || Date.now(),
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply session/index.js]: Corrected.

File: analytics.js (hunk 3 above)
Comment (MoumitaM) [reply analytics.js]: Done.

File: analytics.js (hunk 4 above)
Comment (MoumitaM) [reply analytics.js]: Done.

File: session/index.js [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+    this.sessionInfo = this.storage.getSessionInfo() || {};
+    /**
+     * By default this.autoTrack will be true
+     * Cases where this.autoTrack will be false:
+     * 1. User explicitly set autoTrack load option to false
+     * 2. When user is manually tracking the session
+     *
+     * Depending on the use case, this.autoTrack is set to true/false.
+     */
+    this.autoTrack = !(
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply session/index.js]: Done.

File: session/index.js [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+    if (this.sessionInfo.manuallyTrackSession) return;
+    if (sessionId && typeof sessionId !== 'string') {
+      logger.error(`[Session]:: "sessionId" should be in string format`);
+      return;
+    }
+    this.sessionInfo = {
+      id: sessionId || Date.now().toString(),
+      sessionStart: true,
+      manuallyTrackSession: true,
+      autoTrack: false,
```
Comment (saikumarrs) [inline session/index.js:122]: This is not required as you are resetting the whole object.

File: session/index.js [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+   */
+  end() {
+    this.sessionInfo = {};
+    this.storage.removeSessionInfo();
+  }
+
+  /**
+   * A function get ongoing sessionId.
+   */
+  getSession() {
```
Comment (saikumarrs) [inline session/index.js:138]: `getSessionInfo`

File: session/index.js [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+    const session = {};
```
Comment (saikumarrs) [inline session/index.js:139]: It's better to return `undefined` by default.

File: session/index.js [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+        }
+      }
+      if (this.sessionInfo.sessionStart) {
+        this.sessionInfo.sessionStart = false;
+        session.sessionStart = true;
+      }
+      this.storage.setSessionInfo(this.sessionInfo);
+      session.sessionId = this.sessionInfo.id;
+    }
+    return session;
```
Comment (saikumarrs) [inline session/index.js:155]:
```suggestion
      // renew or create a new auto-tracking session
      if (this.sessionInfo.autoTrack) {
        const timestamp = Date.now();
        if (!this.isValidSession(timestamp)) {
          this.startAutoTracking();
        }
        else {
          this.sessionInfo.expiresAt = timestamp + this.timeout; // set the expiry time of the session
        }
      }
      
      if (this.sessionInfo.sessionStart) {
        session.sessionStart = true;
        this.sessionInfo.sessionStart = false;
      }
      session.sessionId = this.sessionInfo.id;
      
      this.storage.setSessionInfo(this.sessionInfo);
    }
    return session;
```

File: session/index.js [hunk 16]
```diff
@@ -0,0 +1,171 @@
...
+  }
+
+  reset() {
+    const { manuallyTrackSession, autoTrack } = this.sessionInfo;
+    this.sessionInfo = {};
```
Comment (saikumarrs) [inline session/index.js:160]: This is not required if you address one of the comments above.

File: session/index.js [hunk 17]
```diff
@@ -0,0 +1,171 @@
...
+      this.startAutoTracking();
+    } else if (manuallyTrackSession) {
+      this.start();
+    }
+  }
+}
+
+const userSession = new UserSession();
+
+export { userSession as UserSession };
```
Comment (saikumarrs) [inline session/index.js:171]:
```suggestion

export { new UserSession() as UserSession };
```

File: analytics.js [hunk 18]
```diff
@@ -108,6 +109,7 @@ class Analytics {
     this.cookieConsentOptions = {};
     // flag to indicate client integrations` ready status
     this.clientIntegrationsReady = false;
+    this.session = UserSession;
```
Comment (saikumarrs) [inline analytics.js:112]:
```suggestion
    this.uSession = UserSession;
```

**Review by saikumarrs — CHANGES_REQUESTED** [review]

File: session/index.js [hunk 19]
```diff
@@ -0,0 +1,171 @@
...
+    this.storage.setSessionInfo(this.sessionInfo);
+  }
+
+  /**
+   * A public method to start a session
+   * @param {string} sessionId     session identifier
+   * @returns
+   */
+  start(sessionId) {
+    if (this.sessionInfo.manuallyTrackSession) return;
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply session/index.js:113]: Checked with the mobile team, and you are right, they are ending the session and creating a new one. Changing the logic accordingly.

File: session/index.js (hunk 12 above)
Comment (MoumitaM) [reply session/index.js:122]: What is not required?

Comment (saikumarrs) [reply session/index.js:120]: `autoTrack: false`

File: session/index.js (hunk 16 above)
Comment (MoumitaM) [reply session/index.js]: Addressed

File: session/index.js (hunk 15 above)
Comment (MoumitaM) [reply session/index.js]: Addressed

File: session/index.js (hunk 13 above)
Comment (MoumitaM) [reply session/index.js]: Addressed

File: session/index.js [hunk 20]
```diff
@@ -0,0 +1,171 @@
...
+          logger.warn('[Session]:: It is not advised to set "timeout" less than 10 seconds');
+        }
+        this.timeout = timeout;
+      }
+      // If auto session tracking is enabled start the session tracking
+      if (this.sessionInfo.autoTrack) {
+        this.startAutoTracking();
+      } else if (
+        Object.keys(this.sessionInfo).length > 1 &&
+        !this.sessionInfo.manuallyTrackSession
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply session/index.js]: Addressed

File: session/index.js (hunk 19 above)
[the remark this answers is not in the document]
Comment (MoumitaM) [reply session/index.js]: Done

File: session/index.js [hunk 21]
```diff
@@ -0,0 +1,171 @@
...
+   * @returns
+   */
+  start(sessionId) {
+    if (this.sessionInfo.manuallyTrackSession) return;
+    if (sessionId && typeof sessionId !== 'string') {
+      logger.error(`[Session]:: "sessionId" should be in string format`);
+      return;
+    }
+    this.sessionInfo = {
+      id: sessionId || Date.now().toString(),
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply session/index.js]: Addressed

File: session/index.js (hunk 12 above)
Comment (MoumitaM) [reply session/index.js]: Addressed

File: session/index.js (hunk 21 above)
[the remark this answers is not in the document]
Comment (MoumitaM) [reply session/index.js]: it is only applicable for the public method

File: session/index.js [hunk 22]
```diff
@@ -0,0 +1,171 @@
...
+    return timestamp <= this.sessionInfo.expiresAt;
+  }
+
+  /**
+   * A function to check for existing session details and depending on that create a new session.
+   */
+  startAutoTracking() {
+    const timestamp = Date.now();
+    if (!this.isValidSession(timestamp)) {
+      this.sessionInfo.id = timestamp.toString(); // set the current timestamp
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply session/index.js:106]: Addressed

File: session/index.js [hunk 23]
```diff
@@ -0,0 +1,171 @@
...
+      );
+      /**
+       * Validate "timeout" input. Should be provided in milliseconds.
+       * Session timeout: By default, a session lasts until there's 30 minutes of inactivity,
+       * but you can configure this limit using "timeout" load option
+       */
+      if (
+        options &&
+        options.sessions &&
+        typeof options.sessions.timeout === "number"
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply session/index.js]: Addressed

[the remark this answers is not in the document]
Comment (MoumitaM) [reply session/index.js]: Addressed

**Review by saikumarrs — DISMISSED** [review]
LGTM.

**Review by saikumarrs — APPROVED** [review]

**Review by saikumarrs — DISMISSED** [review]

File: session/index.js [hunk 24]
```diff
@@ -111,12 +112,18 @@ class UserSession {
...
    * @returns
    */
   start(sessionId) {
-    if (sessionId && typeof sessionId !== 'string') {
-      logger.error(`[Session]:: "sessionId" should be in string format`);
+    if (sessionId && typeof sessionId !== 'number') {
+      logger.error(`[Session]:: "sessionId" should be in number format`);
+      return;
+    }
+    if (sessionId && countDigits(sessionId) < MIN_SESSION_ID_LENGTH) {
```
Comment (saikumarrs) [inline session/index.js:123]:
```suggestion
    if (countDigits(sessionId) < MIN_SESSION_ID_LENGTH) {
```

Comment (saikumarrs) [reply session/index.js:123]: Redundant.

File: session/index.js [hunk 25]
```diff
@@ -111,12 +112,18 @@ class UserSession {
...
   /**
    * A public method to start a session
-   * @param {string} sessionId     session identifier
+   * @param {number} sessionId     session identifier
    * @returns
    */
   start(sessionId) {
-    if (sessionId && typeof sessionId !== 'string') {
-      logger.error(`[Session]:: "sessionId" should be in string format`);
+    if (sessionId && typeof sessionId !== 'number') {
```
Comment (saikumarrs) [inline session/index.js:119]: What about floating point numbers?

File: session/index.js [hunk 26]
```diff
@@ -111,12 +112,18 @@ class UserSession {
...
+      logger.error(`[Session]:: "sessionId" should be in number format`);
+      return;
+    }
+    if (sessionId && countDigits(sessionId) < MIN_SESSION_ID_LENGTH) {
+      logger.error(
+        `[Session]:: "sessionId" should contain ${MIN_SESSION_ID_LENGTH} or more numbers`
```
Comment (saikumarrs) [inline session/index.js:125]: Incorrect.
```suggestion
        `[Session]:: "sessionId" should at least be "${MIN_SESSION_ID_LENGTH}" digits long`
```

File: session/index.js [hunk 27]
```diff
@@ -111,12 +112,18 @@ class UserSession {
...
    * A public method to start a session
-   * @param {string} sessionId     session identifier
+   * @param {number} sessionId     session identifier
    * @returns
    */
   start(sessionId) {
-    if (sessionId && typeof sessionId !== 'string') {
-      logger.error(`[Session]:: "sessionId" should be in string format`);
+    if (sessionId && typeof sessionId !== 'number') {
+      logger.error(`[Session]:: "sessionId" should be in number format`);
```
Comment (saikumarrs) [inline session/index.js:120]:
```suggestion
      logger.error(`[Session]:: "sessionId" should be a number`);
```

File: utils/utils.js [hunk 28]
```diff
@@ -737,6 +737,10 @@ const constructPayload = (object, mapper) => {
   return payload;
 };
 
+const countDigits = (number) => {
+  return number.toString().length;
```
Comment (saikumarrs) [inline utils/utils.js:741]: Handle failure scenarios.

**Review by saikumarrs — CHANGES_REQUESTED** [review]

File: session/index.js (hunk 25 above)
Comment (saikumarrs) [reply session/index.js]: Truncation is not a good idea.

File: dist/rudder-sdk-js/index.d.ts [hunk 29]
```diff
@@ -86,6 +86,10 @@ declare module "rudder-sdk-js" {
     cookieConsentManager?: cookieConsentManager;
     anonymousIdOptions?: anonymousIdOptions;
     sameSiteCookie?: string;
+    sessions?: {
```
Comment (saikumarrs) [inline dist/rudder-sdk-js/index.d.ts:89]: Session APIs are not added.

Comment (MoumitaM) [reply dist/rudder-sdk-js/index.d.ts:89]: Done

File: dist/rudder-sdk-js/index.d.ts [hunk 30]
```diff
@@ -438,6 +438,16 @@ declare module "rudder-sdk-js" {
...
 
+  /**
+   * To manually start user session in the SDK
+   */
+   function startSession(sessionId?: number): void;
+
+   /**
+   * To manually end user session in the SDK
+   */
+  function endSession(): void;
```
Comment (saikumarrs) [inline dist/rudder-sdk-js/index.d.ts:449]:
```suggestion
   function endSession(): void;
```

File: session/index.js (hunk 25 above)
Comment (MoumitaM) [reply session/index.js]: Addressed

File: session/index.js [hunk 31]
```diff
@@ -117,18 +117,17 @@ class UserSession {
    * @returns
    */
   validateSessionId(sessionId) {
-    if (typeof sessionId !== 'number') {
-      logger.error(`[Session]:: "sessionId" should be a number`);
+    if (typeof sessionId !== 'number' || sessionId % 1 !== 0) {
+      logger.error(`[Session]:: "sessionId" should only contain neumerical digits`);
```
Comment (saikumarrs) [inline session/index.js:121]:
```suggestion
      logger.error(`[Session]:: "sessionId" should only be a positive integer`);
```

**Review by saikumarrs — APPROVED** [review]

**Pull request merged**